src/java/org/apache/cassandra/service/accord/serializers/TableMetadatas.java

```diff
@@ -0,0 +1,283 @@
+        {
+        }
+
+        @Override
+        public void serializeSelf(DataOutputPlus out) throws IOException
+        {
+            out.writeUnsignedVInt32(1);
+            table.id.serializeCompactComparable(out);
+        }
+
+        @Override
+        public TableMetadata deserialize(DataInputPlus in) throws IOException
+        {
+            return table;
+        }
+
+        @Override
+        public long serializedSize(TableMetadata table)
+        {
+            return 0;
+        }
+
+        @Override
+        public long serializedSelfSize()
+        {
+            return TypeSizes.sizeofUnsignedVInt(1) + table.id.serializedCompactComparableSize();
+        }
+
+        @Override
+        public int indexOf(TableId tableId)
+        {
+            if (tableId.equals(table.id))
+                return 0;
+            return -1;
+        }
+
+        @Override
+        public TableMetadata get(TableId tableId)
+        {
+            if (tableId.equals(table.id))
+                return table;
+            return null;
+        }
+    }
+
+    static class Multi extends TableMetadatas
+    {
+        static final TableMetadatas NONE = new Multi();
+
+        final TableMetadata[] tables;
+
+        Multi(TableMetadata ... tables)
+        {
+            this.tables = tables;
+        }
+
+        @Override
+        public TableMetadata get(int index)
+        {
+            return tables[index];
+        }
+
+        @Override
+        public int size()
+        {
+            return tables.length;
+        }
+
+        @Override
+        public int indexOf(TableMetadata find)
+        {
+            return Arrays.binarySearch(tables, find, comparingId);
+        }
+
+        @Override
+        public int indexOf(TableId find)
+        {
+            return SortedArrays.binarySearch(tables, 0, tables.length, find, (id, metadata) -> id.compareTo(metadata.id), FAST);
+        }
+
+        @Override
+        public void serialize(TableMetadata table, DataOutputPlus out) throws IOException
+        {
+            int i = indexOf(table);
+            if (i < 0)
+                throw new IllegalStateException("TableMetadata for " + table + " not found in " + this);
+            out.writeUnsignedVInt32(i);
+        }
+
+        @Override
+        public void serializeSelf(DataOutputPlus out) throws IOException
+        {
+            out.writeUnsignedVInt32(tables.length);
+            for (TableMetadata table : tables)
+                table.id.serializeCompactComparable(out);
+        }
+
+        @Override
+        public TableMetadata deserialize(DataInputPlus in) throws IOException
+        {
+            return get(in.readUnsignedVInt32());
+        }
+
+        @Override
+        public long serializedSize(TableMetadata table)
+        {
+            int i = indexOf(table);
+            if (i < 0)
+                throw new IllegalStateException("TableMetadata for " + table + " not found in " + this);
```

Comment:
this is all duplicate logic, can we refactor into a shared method?

Reply:
You mean just these three lines that are duplicated twice? I'm weakly against introducing a new method for this, but don't mind being overruled.